### ISSUE TITLE
In included file: unexpected OpenMP directive

### ISSUE THREAD

**Issue by havogt** [clangd/clangd#173]
Having an openmp pragma in a header results in the following error shown on the include.
>In included file: unexpected OpenMP directive '#pragma omp parallel for'clang(omp_unexpected_directive)
omp_header.hpp(4, 22): Error occurred here

Tested on the master from yesterday. Problem was already present in clangd 8.0.1 (no diagnostic is displayed, seen in the logs).

compile command `clang++ -fopenmp main.cc`
```cpp
// header.hpp
#pragma once

void loop() {
#pragma omp parallel for
  for (int i = 0; i < 100; ++i) {
  }
}
```

```cpp
// main.cc
#include "header.hpp"

int main() {
    loop();
}
```

The omp pragma seems to mess up the parsing of the header file somehow as I see some follow-up errors in the log:
```
V[14:12:23.340] Dropped diagnostic: /home/vogtha/tmp/omp_problem/main.cc: in included file: expected unqualified-id
V[14:12:23.346] index AST for /home/vogtha/tmp/omp_problem/main.cc (main=false): 
  symbol slab: 1 symbols, 4456 bytes
  ref slab: 0 symbols, 0 refs, 136 bytes
  relations slab: 0 relations, 24 bytes
V[14:12:23.347] Built preamble of size 202656 for file /home/vogtha/tmp/omp_problem/main.cc
V[14:12:23.347] Dropped diagnostic: /home/vogtha/tmp/omp_problem/main.cc: in included file: extraneous closing brace ('}')
V[14:12:23.349] index AST for /home/vogtha/tmp/omp_problem/main.cc (main=true): 
  symbol slab: 1 symbols, 4456 bytes
  ref slab: 1 symbols, 1 refs, 4256 bytes
  relations slab: 0 relations, 24 bytes
```

The problem was also reported here https://bugreports.qt.io/projects/QTCREATORBUG/issues/QTCREATORBUG-22225?filter=allissues (assuming they have clangd activated).

**Comment by ilya-biryukov** [comment]
Could you provide a log line with the compile command?
Maybe it's missing `-fopenmp`...

**Comment by havogt** [comment]
It is there. If I remove `-fopenmp` from the compile_commands.json, no error is displayed as the pragma is properly ignored.

```
Updating file /home/vogtha/tmp/omp_problem/main.cc with command 
[/home/vogtha/tmp/omp_problem]
/usr/bin/g++-7 --driver-mode=g++ -fopenmp -c /home/vogtha/tmp/omp_problem/main.cc -fsyntax-only -resource-dir=<...>/install/lib/clang/10.0.0
```

**Comment by havogt** [comment]
Yes, you are right. The problematic line is https://github.com/llvm/llvm-project/blob/f3ae951c09ef7ce98c62bd0941fb33cbb32f528b/clang/lib/Parse/Parser.cpp#L278
The body should be skipped, but parsing continues at the openmp pragma. Then the pragma is parsed as a pragma on a declaration which makes the `parallel for` invalid.

What's the reason for continuing parsing at openmp pragmas? Should it be kept?

**Comment by havogt** [comment]
It was added here https://github.com/llvm/llvm-project/commit/1d97d2a3d20e1490a72d3275c37ad57df3ede45c

**Comment by havogt** [comment]
cc @alexey-bataev

**Comment by alexey-bataev** [comment]
Unable to reproduce, for me, everything compiles without any problems.
`clang++ -fopenmp main.cc` does not produce any diagnostic

**Comment by havogt** [comment]
This issue is only in clangd, which tries to skip function bodies.

**Comment by alexey-bataev** [comment]
Then maybe clangd must be fixed? No idea, how to reproduce it with clangd.

**Comment by havogt** [comment]
I believe the fix has to be in Parser.cpp, but maybe only clangd is using that code path (via SkipFunctionBody()). I just thought you might have an idea how to improve the handling of the bug you fixed in the commit mentioned above. For skipping function bodies, exiting on an openmp pragma inside of the function looks too aggressive.

**Comment by alexey-bataev** [comment]
Are you skipping the body using `Parser::SkipFunctionBody()`? Or something else?

**Comment by alexey-bataev** [comment]
If you skipping the tokens using `SkipUntil` function, you need to call in a loop, something like `while (!SkipUntil(...))`, you can't just call it once. It returns `true` if the specified token is found and `false` otherwise. You need to call this function unless the return value is `true`.

**Comment by alexey-bataev** [comment]
Plus, maybe, you will need to check for `tok::eof` in case if the function was not terminated and you reached the end of the file.

**Comment by havogt** [comment]
I already tried this code snippet, but got stuck in an infinite loop.
I don't see where `Tok==tok::annot_pragma_omp` should get consumed. I see only two cases before the early exit, both in branches which don't apply.

**Comment by alexey-bataev** [comment]
Must be consumed here https://github.com/llvm/llvm-project/blob/f3ae951c09ef7ce98c62bd0941fb33cbb32f528b/clang/lib/Parse/Parser.cpp#L256, unless your flag is `StopBeforeMatch`. If the flag is `StopBeforeMatch`, call `ConsumeAnyToken()` in loop body, I think.

**Comment by havogt** [comment]
This is inside the branch where we matched the token https://github.com/llvm/llvm-project/blob/f3ae951c09ef7ce98c62bd0941fb33cbb32f528b/clang/lib/Parse/Parser.cpp#L252

**Comment by alexey-bataev** [comment]
Ah, missed this. Then yes, just call `ConsumeAnyToken();` in the body. But add a check for possible `tok::eof`.

**Comment by havogt** [comment]
I did this but now I run into the other problem I was asking earlier:
>couldn't I end up with an unmatched {? (I mean, in the case where the first call to SkipUntil skipped over a {, exited early (return false), then wouldn't the second call return on the wrong }.)

E.g. in the slightly modified case:
```
#pragma once

void loop() {
{ // <- skipped over this in the first call
#pragma omp parallel for
  for (int i = 0; i < 100; ++i) {
  }
} // <- skips until here
}
```

**Comment by alexey-bataev** [comment]
Just add a check for `tok::annot_pragma_openmp` and `tok::annot_pragma_openmp_end` before consuming the tokens.

**Comment by havogt** [comment]
Sorry, I don't get how it is related to my question. From my understanding exiting on openmp pragma just doesn't make sense in the case where we are trying to skip the function body (or more general in the case of `SkipUntil(tok::r_brace)`).

Maybe this is a conflict of interest between the `clangd` and the `clang++` use-case, because in the latter it is used only for trying to recover from invalid code (if I am not mistaken), while in the former we try to skip valid code. Thought, not sure if my understanding is correct...

**Comment by alexey-bataev** [comment]
> From my understanding exiting on openmp pragma just doesn't make sense in the case where we are trying to skip the function body.

`SkipUntil` does not distinguish between skipping function body and normal processing of the OpenMP pragmas. You need special processing of these tokens when trying to skip the function body.

> (or more general in the case of `SkipUntil(tok::r_brace)`).
When we found some bugs during parsing of the OpenMP pragma, we need to skip the rest of the line until `tok::annot_module_end` is found. So, we need to stop when we see `tok::annot_module_end`. But if we skip `tok::annot_module_begin`, we have an unbalanced OpenMP directive begin-end and it leads to troubles when trying to parse the OpenMP pragmas.

**Comment by havogt** [comment]
Still don't get it. Are you suggesting to introduce a special handling in `SkipUntil` in case the call is originating from the `SkipFunctionBody()`? Under the assumption that the function body which we want to skip, won't have one of the problems for which the return on OpenMP pragma was introduced?

**Comment by alexey-bataev** [comment]
> Are you suggesting to introduce a special handling in SkipUntil in case the call is originating from the SkipFunctionBody()?

This is one of the possible solutions. The second one is to use while loop with the checks for OpenMP tokens before calling `ConsumeAnyToken()`.

**Comment by havogt** [comment]
I don't understand the second approach. Where would you put the while loop?

**Comment by alexey-bataev** [comment]
In the SkipFunctionBody() where you skipping the body of the function.

**Comment by havogt** [comment]
But the token in SkipFunctionBody will be the `{` of the function body, the next token can be anything. Only somewhere later in SkipUntil() we might eventually encounter the OpenMP pragma.

**Comment by alexey-bataev** [comment]
Can you try this:
```
while (!SkipUntil(...)) {
  if (Tok.is(tok::annot_openmp_begin) || Tok.is(tok::annot_openmp_end)
    ConsumeAnnotationToken();
}
```
?

**Comment by havogt** [comment]
The issue in this case is that I might have skipped an unmatched `{`. I would have to keep track how many unmatched `{` and then need to match it with the same number of SkipUntil() calls returning `true`.

**Comment by havogt** [comment]
That's what I meant with
> (or more general in the case of `SkipUntil(tok::r_brace))`.

All calls to `SkipUntil(tok::r_brace))` with an OpenMP pragma inside will have this problem, not only the `SkipFunctionBody()`.

**Comment by alexey-bataev** [comment]
It means, that there is the same problem with modules, code completion token, and semicolon with StopAtSemi flag.

**Comment by havogt** [comment]
I’d argue that code completion and semicolon is ok, as there the early exit is driven by flags, I.e. the caller explicitly asks for it. For modules I am not sure, but maybe they cannot appear in parenthesis.

**Comment by alexey-bataev** [comment]
Possible solution is to add a new flag, similar to a flag for semicolon, and exit from SkipUntil for OpenMP if the flag is set. This flag should be set in all cases we parse OpenMP pragmas, otherwise the flag should not be set.

**Comment by havogt** [comment]
That makes sense. I'll try it.

**Comment by zacwitte** [comment]
I'm also experiencing this linting error any time I have `-fopenmp` compile flag enabled and have a `#pragma omp parallel` directive anywhere in the headers included. Just to clarify - is clangd incompatible with OpenMP as of today? Any ETA for a fix?

**Comment by xEnVrE** [comment]
Any news on this?

**Comment by alexey-bataev** [comment]
Fixed in 8035bb4a6573f7d20f17044a68a1405691000525

**Comment by sam-mccall** [comment]
Thanks Alexey! Closing this for now, patches to improve OpenMP (& OpenCL, Cuda, etc) always welcome!
